File: internal/waf/waf.go
// Package waf implements a lightweight Web Application Firewall with a set of
// regex signatures covering SQLi, XSS, RCE, path traversal, and other OWASP
// CRS-lite patterns. Each signature is compiled once at startup and matched
// with a short timeout to stay ReDoS-safe.
package waf

import (
	"context"
	"io"
	"net/http"
	"regexp"
	"strings"
	"time"

	"github.com/foxai/fox-shield/internal/store"
)

// Signature is a single WAF rule.
type Signature struct {
	ID      string
	Category string
	Pattern *regexp.Regexp
}

// Rule is the declarative form used to build signatures.
type Rule struct {
	ID       string
	Category string
	Regex    string
}

// DefaultRules is the v1.0 signature set (20+ patterns).
var DefaultRules = []Rule{
	// SQL Injection
	{ID: "SQLI-001", Category: "sqli", Regex: `(?i)(union\s+select|select\s+.*\s+from|insert\s+into|drop\s+table|delete\s+from|update\s+.*\s+set)`},
	{ID: "SQLI-002", Category: "sqli", Regex: `(?i)('|")\s*(or|and)\s*('|")\s*=\s*('|")`},
	{ID: "SQLI-003", Category: "sqli", Regex: `(?i)(--\s|#|/\*.*\*/)`},
	{ID: "SQLI-004", Category: "sqli", Regex: `(?i)(sleep\s*\(|benchmark\s*\(|waitfor\s+delay|pg_sleep)`},
	{ID: "SQLI-005", Category: "sqli", Regex: `(?i)(information_schema|sys\.tables|mysql\.user|sqlite_master)`},
	{ID: "SQLI-006", Category: "sqli", Regex: `(?i)(0x[0-9a-f]{8,}|char\s*\(\s*\d+\s*\)|concat\s*\()`},
	// XSS
	{ID: "XSS-001", Category: "xss", Regex: `(?i)(<script|</script|javascript:|onerror\s*=|onload\s*=|onclick\s*=)`},
	{ID: "XSS-002", Category: "xss", Regex: `(?i)(<img[^>]*src|document\.cookie|alert\s*\(|eval\s*\()`},
	{ID: "XSS-003", Category: "xss", Regex: `(?i)(<iframe|<object|<embed|<svg[^>]*on)`},
	{ID: "XSS-004", Category: "xss", Regex: `(?i)(&#x[0-9a-f]{2,};|&#\d{2,};|%3c|%3e)`},
	// Remote Code Execution
	{ID: "RCE-001", Category: "rce", Regex: `(?i)(system\s*\(|exec\s*\(|passthru\s*\(|shell_exec\s*\(|popen\s*\()`},
	{ID: "RCE-002", Category: "rce", Regex: `(?i)(\$\{IFS\}|/bin/(sh|bash)|cmd\.exe|powershell\s+-)`},
	{ID: "RCE-003", Category: "rce", Regex: `(?i)(\|\s*(cat|nc|wget|curl|python|perl|php)\b|;\s*(cat|nc|wget|curl)\b)`},
	{ID: "RCE-004", Category: "rce", Regex: `(?i)(base64_decode\s*\(|eval\s*\(\s*\$|assert\s*\(\s*\$)`},
	{ID: "RCE-005", Category: "rce", Regex: `(?i)(\$\{jndi:|log4shell|log4j)`},
	// Path Traversal
	{ID: "TRAV-001", Category: "traversal", Regex: `(\.\./|\.\.\\|%2e%2e|%252e)`},
	{ID: "TRAV-002", Category: "traversal", Regex: `(?i)(/etc/passwd|/etc/shadow|/proc/self|/windows/win\.ini|/boot\.ini)`},
	{ID: "TRAV-003", Category: "traversal", Regex: `(?i)(file://|php://|data://|expect://|gopher://)`},
	// Other / OWASP CRS-lite
	{ID: "GEN-001", Category: "generic", Regex: `(?i)(<\%|%\>|<\?php|<\?xml)`},
	{ID: "GEN-002", Category: "generic", Regex: `(?i)(\x00|%00|\\x00)`},
	{ID: "GEN-003", Category: "generic", Regex: `(?i)(\.env|\.git/config|\.aws/credentials|id_rsa)`},
	{ID: "GEN-004", Category: "generic", Regex: `(?i)(\badmin\b.*\bpassword\b|root\s*:\s*\*|passwd\s*=\s*)`},
}

// WAF holds the compiled signature set.
type WAF struct {
	sigs    []Signature
	timeout time.Duration
	store   store.Store
}

// New compiles the given rules into a WAF. If rules is nil, DefaultRules is
// used. matchTimeout bounds each regex evaluation to stay ReDoS-safe.
func New(rules []Rule, matchTimeout time.Duration, s store.Store) (*WAF, error) {
	if rules == nil {
		rules = DefaultRules
	}
	if matchTimeout <= 0 {
		matchTimeout = 5 * time.Millisecond
	}
	w := &WAF{timeout: matchTimeout, store: s}
	for _, r := range rules {
		re, err := regexp.Compile(r.Regex)
		if err != nil {
			return nil, err
		}
		w.sigs = append(w.sigs, Signature{ID: r.ID, Category: r.Category, Pattern: re})
	}
	return w, nil
}

// Match scans the request target and body for any signature. It returns the
// first matching signature ID and category, or ("", "") if clean.
func (w *WAF) Match(r *http.Request) (id, category string) {
	target := r.Method + " " + r.URL.RequestURI()
	body := readBody(r)
	haystack := target + "\n" + body

	for _, sig := range w.sigs {
		matched := matchWithTimeout(sig.Pattern, haystack, w.timeout)
		if matched {
			return sig.ID, sig.Category
		}
	}
	return "", ""
}

// Middleware wraps a handler. On a signature match it bans the client IP and
// returns 403.
func (w *WAF) Middleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
		id, cat := w.Match(r)
		if id != "" {
			ip := clientIP(r)
			_ = store.BanReason(r.Context(), w.store, ip, "waf:"+id+":"+cat, 60*time.Minute)
			http.Error(rw, "Forbidden", http.StatusForbidden)
			return
		}
		next.ServeHTTP(rw, r)
	})
}

// matchWithTimeout runs a regex match under a deadline.
func matchWithTimeout(re *regexp.Regexp, s string, timeout time.Duration) bool {
	done := make(chan bool, 1)
	go func() {
		done <- re.MatchString(s)
	}()
	select {
	case res := <-done:
		return res
	case <-time.After(timeout):
		return false
	}
}

// readBody reads and restores the request body so downstream handlers can
// still read it.
func readBody(r *http.Request) string {
	if r.Body == nil {
		return ""
	}
	buf := make([]byte, 0, 4096)
	tmp := make([]byte, 1024)
	for {
		n, err := r.Body.Read(tmp)
		buf = append(buf, tmp[:n]...)
		if err != nil {
			break
		}
		if len(buf) > 1<<20 { // cap at 1 MiB
			break
		}
	}
	r.Body = newReplayBody(buf)
	return string(buf)
}

// replayBody lets a consumed body be re-read.
type replayBody struct {
	data []byte
	pos  int
}

func newReplayBody(data []byte) *replayBody { return &replayBody{data: data} }

func (b *replayBody) Read(p []byte) (int, error) {
	if b.pos >= len(b.data) {
		return 0, io.EOF
	}
	n := copy(p, b.data[b.pos:])
	b.pos += n
	return n, nil
}

func (b *replayBody) Close() error { return nil }

func clientIP(r *http.Request) string {
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		if i := strings.IndexByte(xff, ','); i >= 0 {
			return strings.TrimSpace(xff[:i])
		}
		return strings.TrimSpace(xff)
	}
	host := r.RemoteAddr
	if i := strings.LastIndexByte(host, ':'); i >= 0 {
		return host[:i]
	}
	return host
}

File: edge/worker/src/waf.ts
/**
 * fox-shield edge worker — Web Application Firewall.
 *
 * A set of regex signatures covering SQLi, XSS, RCE, path traversal, Log4j and
 * other OWASP CRS-lite patterns. On a match the request is dark-listed (by
 * hash), the client IP is banned, and a 403 "Blocked by WAF" is returned.
 *
 * The signature set mirrors internal/waf/waf.go (DefaultRules).
 */

import type { Store } from './store';
import { addDark, recordOffense, storeRawIp } from './store';
import { vaultEncrypt } from './hash';

export interface WafRule {
  id: string;
  category: string;
  regex: string;
}

export const DEFAULT_WAF_RULES: WafRule[] = [
  // SQL Injection
  { id: 'SQLI-001', category: 'sqli', regex: '(?i)(union\\s+select|select\\s+.*\\s+from|insert\\s+into|drop\\s+table|delete\\s+from|update\\s+.*\\s+set)' },
  { id: 'SQLI-002', category: 'sqli', regex: '(?i)([\'"])\\s*(or|and)\\s*([\'"])\\s*=\\s*([\'"])' },
  { id: 'SQLI-003', category: 'sqli', regex: '(?i)(--\\s|#|/\\*.*\\*/)' },
  { id: 'SQLI-004', category: 'sqli', regex: '(?i)(sleep\\s*\\(|benchmark\\s*\\(|waitfor\\s+delay|pg_sleep)' },
  { id: 'SQLI-005', category: 'sqli', regex: '(?i)(information_schema|sys\\.tables|mysql\\.user|sqlite_master)' },
  { id: 'SQLI-006', category: 'sqli', regex: '(?i)(0x[0-9a-f]{8,}|char\\s*\\(\\s*\\d+\\s*\\)|concat\\s*\\()' },
  { id: 'SQLI-007', category: 'sqli', regex: '(?i)(\\sor\\s+[0-9]+\\s*=\\s*[0-9]+)' },
  // XSS
  { id: 'XSS-001', category: 'xss', regex: '(?i)(<script|</script|javascript:|onerror\\s*=|onload\\s*=|onclick\\s*=)' },
  { id: 'XSS-002', category: 'xss', regex: '(?i)(<img[^>]*src|document\\.cookie|alert\\s*\\()' },
  { id: 'XSS-003', category: 'xss', regex: '(?i)(<iframe|<object|<embed|<svg[^>]*on)' },
  { id: 'XSS-004', category: 'xss', regex: '(?i)(&#x[0-9a-f]{2,};|&#\\d{2,};|%3c|%3e)' },
  // Remote Code Execution
  { id: 'RCE-001', category: 'rce', regex: '(?i)(system\\s*\\(|exec\\s*\\(|passthru\\s*\\(|shell_exec\\s*\\(|popen\\s*\\()' },
  { id: 'RCE-002', category: 'rce', regex: '(?i)(\\$\\{IFS\\}|/bin/(sh|bash)|cmd\\.exe|powershell\\s+-)' },
  { id: 'RCE-003', category: 'rce', regex: '(?i)(\\|\\s*(cat|nc|wget|curl|python|perl|php)\\b|;\\s*(cat|nc|wget|curl)\\b)' },
  { id: 'RCE-004', category: 'rce', regex: '(?i)(base64_decode\\s*\\(|eval\\s*\\(\\s*\\$|assert\\s*\\(\\s*\\$)' },
  { id: 'RCE-005', category: 'rce', regex: '(?i)(\\$\\{jndi:|log4shell|log4j)' },
  // Path Traversal
  { id: 'TRAV-001', category: 'traversal', regex: '(\\.\\./|\\.\\.\\\\|%2e%2e|%252e)' },
  { id: 'TRAV-002', category: 'traversal', regex: '(?i)(/etc/passwd|/etc/shadow|/proc/self|/windows/win\\.ini|/boot\\.ini)' },
  { id: 'TRAV-003', category: 'traversal', regex: '(?i)(file://|php://|data://|expect://|gopher://)' },
  // Other / OWASP CRS-lite
  { id: 'GEN-001', category: 'generic', regex: '(?i)(<%|%>|<\\?php|<\\?xml)' },
  { id: 'GEN-002', category: 'generic', regex: '(?i)(\\x00|%00|\\\\x00)' },
  { id: 'GEN-003', category: 'generic', regex: '(?i)(\\.env|\\.git/config|\\.aws/credentials|id_rsa)' },
  { id: 'GEN-004', category: 'generic', regex: '(?i)(\\badmin\\b.*\\bpassword\\b|root\\s*:\\s*\\*|passwd\\s*=\\s*)' },
];

export interface WafMatch {
  id: string;
  category: string;
  /** True when the request body exceeded the scan cap and could not be fully inspected. */
  oversizedBody: boolean;
}

// maxBodyBytes caps how much of the request body is scanned. Bodies larger
// than this cannot be fully inspected and are treated as suspicious rather
// than silently truncated (which would let a payload hide past the cap).
const maxBodyBytes = 1 << 20; // 1 MiB

// hopByHopHeaders are excluded from header scanning per RFC 7230 §6.1.
const hopByHopHeaders = new Set([
  'connection',
  'keep-alive',
  'proxy-authenticate',
  'proxy-authorization',
  'te',
  'trailer',
  'transfer-encoding',
  'upgrade',
]);

/** Joins relevant request headers (lowercased name:value) into a scan string. */
function headerHaystack(headers: Headers): string {
  const parts: string[] = [];
  headers.forEach((value, name) => {
    const lower = name.toLowerCase();
    if (hopByHopHeaders.has(lower)) {
      return;
    }
    parts.push(`${lower}:${value}`);
  });
  return parts.join('\n');
}

export class Waf {
  private readonly rules: WafRule[];
  private readonly store: Store;

  constructor(store: Store, rules: WafRule[] = DEFAULT_WAF_RULES) {
    this.store = store;
    this.rules = rules;
  }

  /**
   * Scans the request target, headers and body for any signature. Returns the
   * first matching rule, or a result with empty id/category when the body was
   * oversized (treated as suspicious), or null when clean.
   */
  match(
    method: string,
    pathname: string,
    search: string,
    body: string,
    headers?: Headers,
    oversizedBody = false,
  ): WafMatch | null {
    const target = `${method} ${pathname}${search}`;
    const headerStr = headers ? headerHaystack(headers) : '';
    const haystack = `${target}\n${headerStr}\n${body}`;
    for (const rule of this.rules) {
      try {
        // Go/PCRE inline (?i) flag is not valid in JS; translate to the `i` flag.
        let pattern = rule.regex;
        let flags = '';
        if (pattern.startsWith('(?i)')) {
          pattern = pattern.slice(4);
          flags = 'i';
        }
        if (new RegExp(pattern, flags).test(haystack)) {
          return { id: rule.id, category: rule.category, oversizedBody };
        }
      } catch {
        // A malformed rule must never crash the worker; skip it.
        continue;
      }
    }
    if (oversizedBody) {
      return { id: '', category: 'oversized-body', oversizedBody: true };
    }
    return null;
  }

  /**
   * Records a WAF hit: dark-lists the request hash and escalates the offense
   * ladder (see store.recordOffense). The raw IP is never used as a key — only
   * the hashed IP, device hash and subnet hash are. The raw IP is stored
   * encrypted in the vault for admin recovery.
   */
  async block(
    ipHash: string,
    deviceHash: string,
    subnetHash: string | null,
    rawIp: string,
    salt: string,
    hash: string,
    normalized: string,
    match: WafMatch,
    aggressive: boolean,
  ): Promise<void> {
    await addDark(this.store, hash, normalized, 60 * 60);
    const reason = match.id ? `waf:${match.id}:${match.category}` : 'waf:oversized-body';
    await recordOffense(this.store, ipHash, deviceHash, subnetHash, reason, aggressive);
    await storeRawIp(this.store, ipHash, rawIp, salt, vaultEncrypt);
  }
}
